<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <a class="navbar-brand" href="_Dataoving1.ipynb">Dataøving 1</a>
    </div>
    <ul class="nav navbar-nav">
        <li><a href="Oppgave1.ipynb">Oppgave 1 - Fra matematiske uttrykk til python-kode</a></li>
        <li><a href="Oppgave2.ipynb">Oppgave 2 - Plotting av matematiske funksjoner</a></li>
        <li><a href="Oppgave3.ipynb">Oppgave 3 - Plotting av flere datasett</a></li>
        <li class="active"><a href="Oppgave4.ipynb">Oppgave 4 - Analyse av lydsignal</a></li>
        <li><a href="Oppgave5.ipynb">Oppgave 5 - For de som vil ha en ekstra utfordring</a></li>
    </ul>
  </div>
</nav>

# Analyse av Lydsignal

**Læringsmål:**
* Kjenne til sammenhengen mellom samplenummer og samplingstidspunkt
* Kunne tyde grafisk fremstilling av signal i både tid- og frekvensplan
* Bli kjent med konseptet *grunnfrekvens*

**Forarbeid:**
* Ha hodetelefoner tilgjengelig

**Bibliotek og notebook-konfigurasjon:**

In [1]:
from numpy import sin, cos, exp, sqrt, log, pi
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile                # Lydfil I/O
from kildekode_data1 import signalAnalyzer

%matplotlib ipympl

**Importering av lydfil.**

Vi skal nå bruke Python til å analysere et lydklipp med spekkhoggerlyder, [orca.wav](orca.wav). Koden nedenfor laster inn lydsignalet som et numpy ndarray `xn` med flyttall mellom -1,0 og 1,0. Dette er en liste med *alle* samplene i lydklippet. I tillegg henter vi også informasjon om samplingsfrekvensen til lydklippet i variabelen `fs`. 

In [2]:
fs, data = wavfile.read(r'orca.wav') # les audiofilen 'orca.wav' og lagre til variablene 'fs' og 'data'
xn = data/0x8000

## a)
Kjør cellen ovenfor, og undersøk variablene `xn`og `fs`.\
Svar på følgende spørsmål:
* Hva er samplingsfrekvensen $f_s$ til lydsignalet?
* Hva er samplingsperioden $T_s$ til lydsignalet?
* Hvor mange sampler består lydfilen av?
* Hva er lengden til lydfilen i sekund?

* Hva er samplingsfrekvensen $f_s$ til lydsignalet? $f_s = 22050 Hz$
* Hva er samplingsperioden $T_s$ til lydsignalet?   $T_s = \frac{1}{22050 Hz} \approx 45 \mu s$
* Hvor mange sampler består lydfilen av? Antall sampler $N$: `len(xn) = 70411`
* Hva er lengden til lydfilen i sekund?  Lydfil Varighet: $T = \frac{N}{22050} \approx 3.19 s$

**Bruk av funksjonen `signalAnalyzer()`**

Med denne dataøvingen følger en funksjon som genererer en figur med plot av signalinnhold i både tid og frekvens for et bestemt utklipp av et digitalt signal. Denne kan brukes til å velge ut et segment av lydsignalet `xn` og analysere dette i både tids- og frekvensdomenet. Denne skal vi bruke videre i oppgaven til å analysere innholdet lydfilen `orca.wav`.

In [3]:
signalAnalyzer(xn, fs);
# Starter opp et grafisk verktøy for å analysere signalet xn
# Her kan du velge første og siste sample i et "signalvindu" du ønsker 
# å inspisere nærmere. Du kan også velge om verktøyet skal vise signalkurven
# som funksjon av tid, eller signalets frekvensinnhold.

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

Output()

## b)
Vi vil studere et mindre utklipp av signalet i tidsintervallet $2.20s \leq t < 2.30s$ i større detalj. Finn ut hvilket sampleintervall som korresponderer med dette tidsintervallet, og bruk verktøyet `signalAnalyzer()` til å vise en analyse av lydklippet i dette tidsintervallet. Verifiser med figuren at utklippet du har valgt er riktig. Oppgaven besvares med å regne ut de to samplenummerene nedenfor.

P.S. Husk at du kan zoome inn på fuiguren, og lese av x- og y koordinater ved å holde musepekeren over et punkt på grafen.

In [10]:
n_1 = round(2.20*fs)
n_2 = round(2.30*fs)

print(n_1, n_2)

T_0 = (2.21858-2.20527)/5
f_0 = 1/T_0
print(T_0, f_0)

48510 50715
0.002662000000000031 375.6574004507845


## c)
Grunnfrekvensen $f_0$ til et periodisk signal defineres som hvor mange ganger et gitt signalmønster gjentear seg selv i løpet av ett sekund. Dette kan beskrives matematisk med følgende formel:
$$ x(t) = x\left( t-\frac{k}{f_0} \right), \ \ \ k \in \mathbb{Z}$$
Signal hentet fra virkeligheten er sjeldent 100% periodiske, men de kan allikevel ha en dominerende periodisk komponent. 
* Studer tidsplottet av signalsegmentet fra oppgave a, og finn grunnperioden $T_0 = \frac{1}{f_0}$ ved å måle distanse i tid mellom gjentagende signalmøsnter. Du kan lese av x/y verdier for ethvert punkt på en figur ved å holde musepekeren over det aktuelle punktet.
*P.S.: Her kan det være en fordel å måle over et større antall perioder (f.eks. 5) og så dele på 5 for å få en gjennomsnittsmåling. Dette vil redusere målefeil* 

Forslag:

Telte fem hele signalperioder i intervallet $ 2.20527s < t < 2.21858s$. Dette gir:

$$5\cdot T_0 = 2.21858s - 2.20527s$$
$$T_0 \approx 2.66 ms$$ 

# d) 
En annen egenskap som kjennetegner sterkt periodiske signal med en grunnfrekvens $f_0$ er at de er satt sammen primært av frekvenskomponenter med som alle er harmoniske med grunnfrekvensen: $f \in \{1\cdot f_0, \ 2\cdot f_0, \ 3\cdot f_0, \ \ldots \}$. Dette kan vi identifisere ved å studere frekvensspekteret til et signal, ettersom det vil oppstå en serie med toppunkt med jevne frekvensintervall $f_0$.

* Studer plottet av frekvensspekteret til signalet, og finn signalets grunnfrekvens $f_0$.
* Kontroller svaret ved å sammenlikne med lønsingen fra deloppgave **c**.

Forslag:

De tre første toppunktene er: 
$$\{372Hz, 744Hz,  1138Hz, \ldots\}$$

Her er det noe unøyaktighet grunnet litt lav oppløsning i frekvensanalysen, men grunnfrekvensen kan sies å være omtrent 

$$f_0 = 372Hz$$


Utregnet grunnfrekvens med basis i grunnperiode fra oppgave c:
$$ f_0 = \frac{1}{T_0} = \frac{1}{2.66ms} \approx 376 Hz$$

Ikke nøyaktig likt resultatet fra oppgave c, men forskjellen er liten. Tatt i betraktning at frekvensoppløsningen er såppas lav for dette korte utklippet, er det ikke uventet at vi får litt avvik.